cli/commands/benchmark-web-vitals.mjs

```diff
@@ -0,0 +1,216 @@
+/**
+ * CLI command to benchmark several URLs for Core Web Vitals and other key metrics.
+ *
+ * WPP Research, Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import puppeteer from 'puppeteer';
+import round from 'lodash-es/round.js';
+
+/**
+ * Internal dependencies
+ */
+import { getURLs } from '../lib/cli/args.mjs';
+import {
+	log,
+	formats,
+	table,
+	isValidTableFormat,
+	OUTPUT_FORMAT_TABLE,
+} from '../lib/cli/logger.mjs';
+import { calcMedian } from '../lib/util/math.mjs';
+
+export const options = [
+	{
+		argname: '-u, --url <url>',
+		description: 'An URL to run benchmark tests for',
+	},
+	{
+		argname: '-n, --number <number>',
+		description: 'Number of requests to perform',
+		defaults: 1,
+	},
+	{
+		argname: '-f, --file <file>',
+		description: 'File with URLs to run benchmark tests for',
+	},
+	{
+		argname: '-o, --output <output>',
+		description: 'Output format: csv or table',
+		defaults: OUTPUT_FORMAT_TABLE,
+	},
+];
+
+export async function handler( opt ) {
+	if ( ! isValidTableFormat( opt.output ) ) {
+		log(
+			formats.error(
+				'The output format provided via the --output (-o) argument must be either "table" or "csv".'
+			)
+		);
+		return;
+	}
+
+	const { number: amount } = opt;
+	const results = [];
+
+	const browser = await puppeteer.launch();
+
+	for await ( const url of getURLs( opt ) ) {
+		const { completeRequests, metrics } = await benchmarkURL(
+			browser,
+			{
+				url,
+				amount,
+			}
+		);
+
+		results.push( [ url, completeRequests, metrics ] );
+	}
+
+	await browser.close();
+
+	if ( results.length === 0 ) {
+		log(
+			formats.error(
+				'You need to provide a URL to benchmark via the --url (-u) argument, or a file with multiple URLs via the --file (-f) argument.'
+			)
+		);
+	} else {
+		outputResults( opt, results );
+	}
+}
+
+async function benchmarkURL( browser, params ) {
+	/*
+	 * For now this only includes load time metrics.
+	 * In the future, additional Web Vitals like CLS, FID, and INP should be
+	 * added, however they are slightly more complex to retrieve through an
+	 * automated headless browser test.
+	 * See https://github.com/GoogleChromeLabs/wpp-research/pull/41.
```

Comment:
FID and INP are more complex as require an interaction, but the load CLS should be similar to FCP/LCP/TTFB shouldn't it?

Granted, that often isn't the full CLS of a page (as it can be impacted by scroll, or other interactions), but it's still usually a good part of CLS. And that "load CLS" is what Lighthouse and PSI already captures despite it's limitations.

The only thing is that, kind of like LCP, CLS is not emitted by web-vitals.js, until the page is hidden (including navigating away from it). For the web-vitals.js library we dummy this visibility change:
https://github.com/GoogleChrome/web-vitals/blob/main/test/utils/stubVisibilityChange.js

And then call it like this:
```js
    // Wait until all images are loaded and rendered, then change to hidden.
    await imagesPainted();
    await stubVisibilityChange('hidden');
```

Maybe that's enough to make it too complicated here?

The alternative is to use the `reportAllChanges` flag and only report at the end of the test.

Reply:
Yeah sounds like it’s due to not changing the visibility so CLS not being “finalised”. Which I realised and edited after writing that first sentence (but before submitting the review).